### ISSUE TITLE
Overridden method Named Arg defaults not being honored.

### ISSUE THREAD

**Issue by glitch** [chapel-lang/chapel#17649]
### Summary of Problem
This is an issue involving inheritance, overriding methods and named argument defaults values; a more complete description of the problem is below.  In short the default values in the overridden method are not being honored, the parent class's are still being used.  I'm not sure if this is expected behavior or a bug; if it's expected maybe the Named Arguments and Default Values documentation could be clarified (https://chapel-lang.org/docs/language/spec/procedures.html#named-arguments)

_**What behavior did you observe when encountering this issue?**_
I declared a base class Foo with a proc `hello` and an inheriting class Bar which overrides `hello`.  In proc `hello` I have named arguments with default values.  Foo has one set of defaults and Bar has another.  When I declare and create a Foo and test out the defaults I get Foo's defaults; when i declare Bar and construct a Bar and repeat the test I get Bar's defaults; so far so good, everything works as expected.  The issue I run into is when I declare a Foo but provide Bar as the concrete implementation.  When I invoke the `hello` proc, it runs the overridden `Bar.hello()` but it gets Foo's defaults.

_**What behavior did you expect to observe?**_
When declaring a Foo and providing Bar as the concrete impl. I expected to get Bar's default values for the named arguments since Bar's `hello` proc was the one that got invoked.

This may be working as the Chapel team intended but it was unexpected and the documentation on Named Arguments and Default Values (https://chapel-lang.org/docs/language/spec/procedures.html#named-arguments) didn't really mention anything about what happens in the case of inheritance.

_**Is this a blocking issue with no known work-arounds?**_
This is not a blocking issue, it's really easy to work around, it's more that it was unexpected behavior, see previous note about possibly intended behavior.

### Steps to Reproduce

**Source Code:**
Note, I also made a gist of it here: https://gist.github.com/glitch/ef9aa1f6643d4dd1429f8d7d4768906d

```chapel
/*
Chapel code example to illustrate issue with inheritance, overriding methods and named vars.
Declare a base class Foo with a hello proc.  It has 3 named args with defaults.
Next Declare Bar which inherits from Foo and overrides Foo's hello proc with its own
default values for the named args.
When we declare a Foo with a concrete implementation of Bar we see that the Bar hello proc
gets invoked as expected, but the default args do not; they are still Foo's defaults.
Since the overridden proc of Bar's gets invoked, I would have expected Bar's defaults to go
along with it, but that's not the case based on the output.
*/
class Foo {

    proc hello(greeting:string="foo default greeting", prefix:string="FOO_default_prefix", suffix:string="FOO_default_suffix"): string throws {
        return "FOO.hello(...)=> greeting:`" + greeting + "`, prefix:`" + prefix + "`,  suffix:`" + suffix + "`";
    }

}

class Bar : Foo {

    override proc hello(greeting:string="bar default greeting", prefix:string = "default_bar[", suffix:string = "]bar_default"): string throws {
        return "I'm in BAR.hello(...)=> greeting:`" + greeting + "`, prefix:`" + prefix + "`,  suffix:`" + suffix + "`";
    }
}


// Create Foo with concrete Foo impl
// 1. Prove we get the defaults
// 2. Pass in greeting and get defaults on prefix & suffix
// 3. Pass in greeting & prefix, get default on suffix
var f:Foo = new Foo();
writeln("----Foo----");
writeln(f.hello());
writeln(f.hello(greeting="hello"));
writeln(f.hello(greeting="hello", prefix="<<<"));

// Create Bar with concrete Bar impl
// 1. Prove we get Bar's defaults
// 2. Pass in greeting and get Bar's defaults on prefix & suffix
// 3. Pass in greeting & prefix, get Bar's default on suffix
var b:Bar = new Bar();
writeln("----Bar----");
writeln(b.hello());
writeln(b.hello(greeting="I'm Bar"));
writeln(b.hello(greeting="I'm Bar", prefix="<<<"));

// Create Foo with conreate Bar impl.
// This uses Bar's overridden proc hello, but doesn't get the Bar default named args, why?
// 1. We see it inokes Bar's hello method in the output, but we get Foo's named defaults
// 2. Pass in greeting, we see Bar's method invoked, but still get Foo defaults on prefix & suffix
// 3. Pass in greetin & suffix, same thing, We see Bar's output message, but still Foo default suffix.
var fb:Foo = new Bar();
writeln("----Foo with Bar as impl----");
writeln(fb.hello());
writeln(fb.hello(greeting="expect Bar defaults"));
writeln(fb.hello(greeting="expect Bar defaults minus passed in prefix", prefix="<<<"));
```

EDIT/UPDATE - Output of sample program:
```
----Foo----
FOO.hello(...)=> greeting:`foo default greeting`, prefix:`FOO_default_prefix`,  suffix:`FOO_default_suffix`
FOO.hello(...)=> greeting:`hello`, prefix:`FOO_default_prefix`,  suffix:`FOO_default_suffix`
FOO.hello(...)=> greeting:`hello`, prefix:`<<<`,  suffix:`FOO_default_suffix`
----Bar----
I'm in BAR.hello(...)=> greeting:`bar default greeting`, prefix:`default_bar[`,  suffix:`]bar_default`
I'm in BAR.hello(...)=> greeting:`I'm Bar`, prefix:`default_bar[`,  suffix:`]bar_default`
I'm in BAR.hello(...)=> greeting:`I'm Bar`, prefix:`<<<`,  suffix:`]bar_default`
----Foo with Bar as impl----
I'm in BAR.hello(...)=> greeting:`foo default greeting`, prefix:`FOO_default_prefix`,  suffix:`FOO_default_suffix`
I'm in BAR.hello(...)=> greeting:`expect Bar defaults`, prefix:`FOO_default_prefix`,  suffix:`FOO_default_suffix`
I'm in BAR.hello(...)=> greeting:`expect Bar defaults minus passed in prefix`, prefix:`<<<`,  suffix:`FOO_default_suffix`
```

**Compile command:**
`chpl default-args-example.chpl `

**Execution command:**
`./default-args-example -nl 1`

### Configuration Information

- Output of `chpl --version`: chpl version 1.24.1
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: none
CHPL_HWLOC: bundled
CHPL_REGEXP: re2
CHPL_LLVM: none
CHPL_AUX_FILESYS: none

- Back-end compiler and version, e.g. `gcc --version` or `clang --version`: gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0

Thanks Chapel Team!

**Comment by bradcray** [comment]
@glitch : Thanks for reporting this.  With a quick look, I agree with you that this behavior seems incorrect (or at least, very surprising), and am finding this issue which looks similar:  https://github.com/chapel-lang/chapel/issues/7880

@mppf: I think you've done the most with default arguments and virtual dispatch lately.  Have you ever looked into how easy or hard this would be to fix?

**Comment by cassella** [comment]
BTW, from that old issue and the ones linked through it, it seems this
misbehavior happens only when the default arguments are `param`s, such
as the string literals in this example.  You might be able to work
around the problem by making variables to use as the default values,

```chapel

const foo_default_greet = "foo default greeting";
class Foo {
    proc hello(greeting:string=foo_default_greet, ...
    ...
}

const bar_default_greet = "bar default greeting";
class Bar : Foo {
    override proc hello(greeting:string=bar_default_greet, ...
    ...
}
```

**Comment by bradcray** [comment]
Huh, that's interesting, Paul.  Though I'd seen the issue about param initializers, I thought it was specific to param functions for some reason... 

That makes me wonder whether @mppf has any follow-ups to our conversation this morning about why this would be.  Specifically, the following rewrite to the case in the OP of #7880 works, which surprises me:

```chapel
var x10 = 10;
var x20 = 20;

class C {
  proc foo(x = x10) { writeln("C.foo:", x); }
}

class D : C {
  override proc foo(x = x20) { writeln("D.foo:", x); }
}

var d = new D();
d.foo();

var c: C = new D();
c.foo();
```

**Comment by mppf** [comment]
@bradcray - the compiler is inferring the default initialization function as being `param` because it is a `param` expression. It should not do that in this case, because we can't virtually dispatch a `param` returning function.

This choice (when is it param?) is something that has been a bit of a challenge over time. Look for the comment "The following code is tricky" in buildDefaultedActualFn. I suspect it would suffice to not set `wrapper->retTag = RET_PARAM` when it's a class method, but I don't know if that presents problems in any cases. I can't remember if I made the code aggressive about setting it to return `param` when possible to address a correctness issue or if I was just trying to avoid adding runtime overhead when it wasn't necessary. I suspect it was the 2nd of these.

**Comment by bradcray** [comment]
@mppf: Thanks for that pointer.  I have a simple change that passes the test from #7880 and the three tricky cases mentioned in the comment.  Now I just need my other test run to complete so I can run it through testing as a whole.

Interestingly, it seems that the implication is that we were generating AST that would do dynamic dispatch even though we're naming a specific FnSymbol in the CallExpr (where we were thinking that was the cause of the static dispatch when we were chatting earlier today).

**Comment by glitch** [comment]
@bradcray No problem and thanks for the quick turn around!

**Comment by bradcray** [comment]
@glitch: I'm sorry to report this, but in taking a victory lap on this fix today, I found that it doesn't work as consistently as I'd thought last week.  Specifically, if we remove all of the cases that statically dispatch to bar (the `b.hello()` cases), we get the wrong defaults being used again.  I've forked off a new issue for this here:  https://github.com/chapel-lang/chapel/issues/17753 and am sorry for jumping the gun in thinking this was fully resolved.
